src/sagemaker/utils.py

```diff
@@ -311,60 +314,90 @@ def repack_model(inference_script, source_directory, model_uri, sagemaker_sessio
         model_uri (str): S3 or file system location of the original model tar
         sagemaker_session (:class:`sagemaker.session.Session`): a sagemaker session to interact with S3.
 
     Returns:
         str: path to the new packed model
     """
-    new_model_name = 'model-%s.tar.gz' % sagemaker.utils.sagemaker_short_timestamp()
+    dependencies = dependencies or []
 
     with _tmpdir() as tmp:
-        tmp_model_dir = os.path.join(tmp, 'model')
-        os.mkdir(tmp_model_dir)
+        model_dir = _extract_model(model_uri, sagemaker_session, tmp)
 
-        model_from_s3 = model_uri.lower().startswith('s3://')
-        if model_from_s3:
-            local_model_path = os.path.join(tmp, 'tar_file')
-            download_file_from_url(model_uri, local_model_path, sagemaker_session)
+        _update_code(model_dir, inference_script, source_directory, dependencies, sagemaker_session, tmp)
 
-            new_model_path = os.path.join(tmp, new_model_name)
-        else:
-            local_model_path = model_uri.replace('file://', '')
-            new_model_path = os.path.join(os.path.dirname(local_model_path), new_model_name)
+        tmp_model_path = os.path.join(tmp, 'temp-model.tar.gz')
+        with tarfile.open(tmp_model_path, mode='w:gz') as t:
+            t.add(model_dir, arcname=os.path.sep)
 
-        with tarfile.open(name=local_model_path, mode='r:gz') as t:
-            t.extractall(path=tmp_model_dir)
+        _save_model(repacked_model_uri, tmp_model_path, sagemaker_session)
 
-        code_dir = os.path.join(tmp_model_dir, 'code')
-        if os.path.exists(code_dir):
-            shutil.rmtree(code_dir, ignore_errors=True)
 
-        if source_directory and source_directory.lower().startswith('s3://'):
-            local_code_path = os.path.join(tmp, 'local_code.tar.gz')
-            download_file_from_url(source_directory, local_code_path, sagemaker_session)
+def _save_model(repacked_model_uri, tmp_model_path, sagemaker_session):
+    if repacked_model_uri.lower().startswith('s3://'):
+        url = parse.urlparse(repacked_model_uri)
+        bucket, key = url.netloc, url.path.lstrip('/')
+        new_key = key.replace(os.path.basename(key), os.path.basename(repacked_model_uri))
 
-            with tarfile.open(name=local_code_path, mode='r:gz') as t:
-                t.extractall(path=code_dir)
+        sagemaker_session.boto_session.resource('s3').Object(bucket, new_key).upload_file(
+            tmp_model_path)
+    else:
+        shutil.move(tmp_model_path, repacked_model_uri.replace('file://', ''))
 
-        elif source_directory:
-            shutil.copytree(source_directory, code_dir)
-        else:
-            os.mkdir(code_dir)
-            shutil.copy2(inference_script, code_dir)
 
-        with tarfile.open(new_model_path, mode='w:gz') as t:
-            t.add(tmp_model_dir, arcname=os.path.sep)
+def _update_code(model_dir, inference_script, source_directory, dependencies, sagemaker_session, tmp):
```

Comment:
this method name doesn't seem entirely indicative of what is going on.

I don't have any suggestions, but this method is mainly copying files to a certain directory and not necessarily updating 'code'.